File: LeetCode/141 - Linked List Cycle/solve.cc
#include <iostream>

#include "solve.h"


int main() {
    {
        ListNode* head = new ListNode(1);
        std::cout << (Solution().hasCycle(head)) << std::endl;
    }
    {
        ListNode* head = new ListNode(1);
        head->next = head;
        std::cout << (Solution().hasCycle(head)) << std::endl;
    }
    {
        ListNode* head = new ListNode(1);
        head->next = new ListNode(2);
        std::cout << (Solution().hasCycle(head)) << std::endl;
    }
    {
        ListNode* head = new ListNode(1);
        head->next = new ListNode(2);
        head->next->next = new ListNode(3);
        std::cout << (Solution().hasCycle(head)) << std::endl;
    }
    {
        ListNode* head = new ListNode(1);
        ListNode* p2 = head->next = new ListNode(2);
        head->next->next = new ListNode(3);
        head->next->next->next = p2;
        std::cout << (Solution().hasCycle(head)) << std::endl;
    }
    return 0;
}

File: LeetCode/141 - Linked List Cycle/solve.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head == None:
            return False
        
        p1 = head
        p2 = head.next
        
        while p1 != p2:
            if p2 == None or p2.next == None:
                return False
            p1 = p1.next
            p2 = p2.next.next
            
        return True


a = ListNode(1)
b = ListNode(2)
a.next = b

s = Solution()
print(s.hasCycle(a))
c = ListNode(3)
d = ListNode(4)
c.next = d
d.next = c
print(s.hasCycle(c))
print(s.hasCycle(d))
